hudi-common/src/test/java/org/apache/hudi/common/table/read/TestSchemaHandler.java

```diff
@@ -0,0 +1,464 @@
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class TestSchemaHandler {
+
+  protected static final Schema DATA_SCHEMA = HoodieAvroUtils.addMetadataFields(HoodieTestDataGenerator.AVRO_SCHEMA);
+  protected static final Schema DATA_COLS_ONLY_SCHEMA = generateProjectionSchema("begin_lat", "tip_history", "rider");
+  protected static final Schema META_COLS_ONLY_SCHEMA = generateProjectionSchema("_hoodie_commit_seqno", "_hoodie_record_key");
+
+  @Test
+  public void testCow() {
+    HoodieReaderContext<String> readerContext = new MockReaderContext(false);
+    readerContext.setHasLogFiles(false);
+    readerContext.setHasBootstrapBaseFile(false);
+    readerContext.setShouldMergeUseRecordPosition(false);
+    HoodieTableConfig hoodieTableConfig = mock(HoodieTableConfig.class);
+    Schema requestedSchema = DATA_SCHEMA;
+    HoodieFileGroupReaderSchemaHandler schemaHandler = new HoodieFileGroupReaderSchemaHandler(readerContext, DATA_SCHEMA,
+        requestedSchema, Option.empty(), hoodieTableConfig, new TypedProperties());
+    assertEquals(requestedSchema, schemaHandler.getRequiredSchema());
+
+    //read subset of columns
+    requestedSchema = generateProjectionSchema("begin_lat", "tip_history", "rider");
+    schemaHandler =
+        new HoodieFileGroupReaderSchemaHandler(readerContext, DATA_SCHEMA, requestedSchema,
+            Option.empty(), hoodieTableConfig, new TypedProperties());
+    assertEquals(requestedSchema, schemaHandler.getRequiredSchema());
+  }
+
+  @Test
+  public void testCowBootstrap() {
+    HoodieReaderContext<String> readerContext = new MockReaderContext(false);
+    readerContext.setHasLogFiles(false);
+    readerContext.setHasBootstrapBaseFile(true);
+    readerContext.setShouldMergeUseRecordPosition(false);
+    HoodieTableConfig hoodieTableConfig = mock(HoodieTableConfig.class);
+    Schema requestedSchema = generateProjectionSchema("begin_lat", "tip_history", "_hoodie_record_key", "rider");
+    HoodieFileGroupReaderSchemaHandler schemaHandler =
+        new HoodieFileGroupReaderSchemaHandler(readerContext, DATA_SCHEMA, requestedSchema,
+            Option.empty(), hoodieTableConfig, new TypedProperties());
+    //meta cols must go first in the required schema
+    Schema expectedRequiredSchema = generateProjectionSchema("_hoodie_record_key", "begin_lat", "tip_history", "rider");
+    assertEquals(expectedRequiredSchema, schemaHandler.getRequiredSchema());
+    Pair<List<Schema.Field>, List<Schema.Field>> bootstrapFields = schemaHandler.getBootstrapRequiredFields();
+    assertEquals(Collections.singletonList(getField("_hoodie_record_key")), bootstrapFields.getLeft());
+    assertEquals(Arrays.asList(getField("begin_lat"), getField("tip_history"), getField("rider")), bootstrapFields.getRight());
+  }
+
+  @Test
+  public void testCowBootstrapWithPositionMerge() {
+    HoodieReaderContext<String> readerContext = new MockReaderContext(true);
```

Comment:
I see lot of code duplications across. atleast all 3 COW related tests, we can dedup.